In [0]:
#import library yang dibutuhkan 

import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

## Import Data

In [0]:
!wget http://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx

--2019-08-09 08:40:46--  http://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 23715344 (23M) [application/x-httpd-php]
Saving to: ‘Online Retail.xlsx.1’

Online Retail.xlsx. 100%[===================>]  22.62M  6.27MB/s    in 3.9s    

2019-08-09 08:40:50 (5.85 MB/s) - ‘Online Retail.xlsx.1’ saved [23715344/23715344]



In [0]:
df = pd.read_excel('Online Retail.xlsx')
df.head(20)

InvoiceNo StockCode  ... CustomerID         Country
0     536365    85123A  ...    17850.0  United Kingdom
1     536365     71053  ...    17850.0  United Kingdom
2     536365    84406B  ...    17850.0  United Kingdom
3     536365    84029G  ...    17850.0  United Kingdom
4     536365    84029E  ...    17850.0  United Kingdom
5     536365     22752  ...    17850.0  United Kingdom
6     536365     21730  ...    17850.0  United Kingdom
7     536366     22633  ...    17850.0  United Kingdom
8     536366     22632  ...    17850.0  United Kingdom
9     536367     84879  ...    13047.0  United Kingdom
10    536367     22745  ...    13047.0  United Kingdom
11    536367     22748  ...    13047.0  United Kingdom
12    536367     22749  ...    13047.0  United Kingdom
13    536367     22310  ...    13047.0  United Kingdom
14    536367     84969  ...    13047.0  United Kingdom
15    536367     22623  ...    13047.0  United Kingdom
16    536367     22622  ...    13047.0  United Kingdom
17    536367     21754  ...    13047.0  United Kingdom
18    536367     21755  ...    13047.0  United Kingdom
19    536367     21777  ...    13047.0  United Kingdom

[20 rows x 8 columns]

In [0]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

## Cleaning Data

In [0]:
#data pada kolom Description yang mengandung strip diawal dan diakhir dihapus
df['Description'] = df['Description'].str.strip()
# data pada kolom InvoiceNo yang mengandung nilai NA maka akan dihapus
df.dropna(axis=0, subset=['InvoiceNo'], inplace=True) 
# data pada kolom InvoiceNo diubah menjadi string
df['InvoiceNo'] = df['InvoiceNo'].astype('str') 
# data pada kolom InvoiceNo yang mengandung karakter C maka akan dihapus
df = df[~df['InvoiceNo'].str.contains('C')]
# tampilkan data
df.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

In [0]:
datanegara = df.groupby(['Country']).count()
datanegara.sort_values(by='Quantity', ascending=False)

InvoiceNo  StockCode  ...  UnitPrice  CustomerID
Country                                     ...                       
United Kingdom           487622     487622  ...     487622      354345
Germany                    9042       9042  ...       9042        9042
France                     8408       8408  ...       8408        8342
EIRE                       7894       7894  ...       7894        7238
Spain                      2485       2485  ...       2485        2485
Netherlands                2363       2363  ...       2363        2363
Belgium                    2031       2031  ...       2031        2031
Switzerland                1967       1967  ...       1967        1842
Portugal                   1501       1501  ...       1501        1462
Australia                  1185       1185  ...       1185        1185
Norway                     1072       1072  ...       1072        1072
Italy                       758        758  ...        758         758
Channel Islands             748        748  ...        748         748
Finland                     685        685  ...        685         685
Cyprus                      614        614  ...        614         614
Sweden                      451        451  ...        451         451
Unspecified                 446        446  ...        446         244
Austria                     398        398  ...        398         398
Denmark                     380        380  ...        380         380
Poland                      330        330  ...        330         330
Japan                       321        321  ...        321         321
Israel                      295        295  ...        295         248
Hong Kong                   284        284  ...        284           0
Singapore                   222        222  ...        222         222
Iceland                     182        182  ...        182         182
USA                         179        179  ...        179         179
Canada                      151        151  ...        151         151
Greece                      145        145  ...        145         145
Malta                       112        112  ...        112         112
United Arab Emirates         68         68  ...         68          68
European Community           60         60  ...         60          60
RSA                          58         58  ...         58          58
Lebanon                      45         45  ...         45          45
Lithuania                    35         35  ...         35          35
Brazil                       32         32  ...         32          32
Czech Republic               25         25  ...         25          25
Bahrain                      18         18  ...         18          17
Saudi Arabia                  9          9  ...          9           9

[38 rows x 7 columns]

## Filtering Data

In [0]:
basket = (df[df['Country'] =="France"]           
          .groupby(['InvoiceNo', 'Description'])['Quantity']           
          .sum().unstack().reset_index().fillna(0)           
          .set_index('InvoiceNo'))
basket.head()

Description  10 COLOUR SPACEBOY PEN  ...  ZINC T-LIGHT HOLDER STARS SMALL
InvoiceNo                            ...                                 
536370                          0.0  ...                              0.0
536852                          0.0  ...                              0.0
536974                          0.0  ...                              0.0
537065                          0.0  ...                              0.0
537463                          0.0  ...                              0.0

[5 rows x 1563 columns]

In [0]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1
basket_sets = basket.applymap(encode_units)
 
#menghapus kolom postage
basket_sets.drop('POSTAGE', inplace=True, axis=1)
basket_sets.head()

Description  10 COLOUR SPACEBOY PEN  ...  ZINC T-LIGHT HOLDER STARS SMALL
InvoiceNo                            ...                                 
536370                            0  ...                                0
536852                            0  ...                                0
536974                            0  ...                                0
537065                            0  ...                                0
537463                            0  ...                                0

[5 rows x 1562 columns]

In [0]:
# Build up the frequent items 
frequent_itemsets = apriori(basket_sets, min_support=0.07, use_colnames=True)
frequent_itemsets.head()

support                        itemsets
0  0.071429   (4 TRADITIONAL SPINNING TOPS)
1  0.096939    (ALARM CLOCK BAKELIKE GREEN)
2  0.102041     (ALARM CLOCK BAKELIKE PINK)
3  0.094388      (ALARM CLOCK BAKELIKE RED)
4  0.081633  (BAKING SET 9 PIECE RETROSPOT)

In [0]:
# Create the rules
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules

antecedents  ... conviction
0                         (ALARM CLOCK BAKELIKE PINK)  ...   3.283859
1                        (ALARM CLOCK BAKELIKE GREEN)  ...   3.791383
2                        (ALARM CLOCK BAKELIKE GREEN)  ...   4.916181
3                          (ALARM CLOCK BAKELIKE RED)  ...   5.568878
4                         (ALARM CLOCK BAKELIKE PINK)  ...   3.293135
5                          (ALARM CLOCK BAKELIKE RED)  ...   4.153061
6                              (DOLLY GIRL LUNCH BOX)  ...   3.102273
7                                (SPACEBOY LUNCH BOX)  ...   2.101190
8                     (PLASTERS IN TIN CIRCUS PARADE)  ...   1.835747
9                          (PLASTERS IN TIN SPACEBOY)  ...   2.363588
10                 (PLASTERS IN TIN WOODLAND ANIMALS)  ...   2.063681
11                    (PLASTERS IN TIN CIRCUS PARADE)  ...   2.104592
12                 (PLASTERS IN TIN WOODLAND ANIMALS)  ...   2.221939
13                         (PLASTERS IN TIN SPACEBOY)  ...   3.443878
14               (SET/20 RED RETROSPOT PAPER NAPKINS)  ...   3.736395
15                      (SET/6 RED SPOTTY PAPER CUPS)  ...   3.345481
16               (SET/20 RED RETROSPOT PAPER NAPKINS)  ...   3.780612
17                    (SET/6 RED SPOTTY PAPER PLATES)  ...   4.336735
18                      (SET/6 RED SPOTTY PAPER CUPS)  ...   7.852041
19                    (SET/6 RED SPOTTY PAPER PLATES)  ...  21.556122
20  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...  ...  34.897959
21  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...  ...  34.489796
22  (SET/6 RED SPOTTY PAPER CUPS, SET/6 RED SPOTTY...  ...   4.625850
23               (SET/20 RED RETROSPOT PAPER NAPKINS)  ...   3.510204
24                      (SET/6 RED SPOTTY PAPER CUPS)  ...   3.232653
25                    (SET/6 RED SPOTTY PAPER PLATES)  ...   4.081633

[26 rows x 9 columns]

In [0]:
filtered_rules = rules[ (rules['lift'] >= 6) &
       (rules['confidence'] >= 0.8) ]

filtered_rules.sort_values(by=['support','confidence'], ascending=False)

antecedents  ... conviction
19                    (SET/6 RED SPOTTY PAPER PLATES)  ...  21.556122
18                      (SET/6 RED SPOTTY PAPER CUPS)  ...   7.852041
17                    (SET/6 RED SPOTTY PAPER PLATES)  ...   4.336735
20  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...  ...  34.897959
21  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...  ...  34.489796
22  (SET/6 RED SPOTTY PAPER CUPS, SET/6 RED SPOTTY...  ...   4.625850
3                          (ALARM CLOCK BAKELIKE RED)  ...   5.568878
2                        (ALARM CLOCK BAKELIKE GREEN)  ...   4.916181

[8 rows x 9 columns]

**Jika Anda bertugas di strategi pemasaran, promo atau paket apa yang Anda tawarkan berdasarkan hasil rule apriori diatas? Jelaskan alasan Anda**

Data filtered_rules telah dilakukan sorting dari nilai support dan confidence yang tertinggi ke terendah.

Support merupakan nilai penunjang yaitu persentase kombinasi item dalam database dan confidence merupakan nilai kepastian yaitu kuatnya hubungan antar item dalam aturan asosiatif. 

{(SET/6 RED SPOTTY PAPER PLATES), 	(SET/6 RED SPOTTY PAPER CUPS)} -> {support = 0.12 atau 12 % , confidence = 0.96 atau 96 %}. 

Pernyataan diatas merupakan nilai support dan nilai confidence tertinggi pada item SET/6 RED SPOTTY PAPER PLATES dan SET/6 RED SPOTTY PAPER CUPS di database. 

Artinya konsumen membeli item SET/6 RED SPOTTY PAPER PLATES  dengan kemungkinan 12% dan memiliki kemungkinan membeli SET/6 RED SPOTTY PAPER PLATES lalu juga membeli  SET/6 RED SPOTTY PAPER CUPS dengan kemungkinan 96 %. 

Sehingga berdasarkan pernyataan tersebut, dapat dilakukan promo jika beli 1 paket yang berisi  2 SET/6 RED SPOTTY PAPER PLATES dan  2 SET/6 RED SPOTTY PAPER CUPS  akan mendapatkan promo berupa diskon. 
